File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        // Sort the array with a custom lambda comparator function
        Arrays.sort(points, (a, b) -> squaredDistance(a) - squaredDistance(b));
        
        // Return the first k elements of the sorted array
        return Arrays.copyOf(points, k);
    }
    
    private int squaredDistance(int[] point) {
        // Calculate and return the squared Euclidean distance
        return point[0] * point[0] + point[1] * point[1];
    }
};


/*
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> maxHeap = new PriorityQueue<int[]>((a,b) -> (a[0]*a[0] + a[1]*a[1]) - (b[0]*b[0] + b[1]*b[1]));

        int[][] ans = new int[k][2];

        for(int i = 0; i < points.length; i++){
            maxHeap.offer(points[i]);
        }

        for(int i = 0; i < k; i++){
            ans[i] = maxHeap.poll();
        }

        return ans;
    }
}
*/

File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        // Sort the 'points' array using a custom comparator
        sort(points.begin(), points.end(), [](const vector<int>& pointA, const vector<int>& pointB) {
            // Calculate the squared Euclidean distance from the origin for pointA
            int distA = pointA[0] * pointA[0] + pointA[1] * pointA[1];
            // Calculate the squared Euclidean distance from the origin for pointB
            int distB = pointB[0] * pointB[0] + pointB[1] * pointB[1];
            // Compare the squared distances to sort the points by distance
            return distA < distB;
        });

        // Construct a vector containing the first 'k' elements from the sorted 'points' array
        return vector<vector<int>>(points.begin(), points.begin() + k);
    }
};
